### PR TITLE
Add support for histograms in WAL checkpointing

### PR DESCRIPTION
**Pull request by codesome**
NOTE: this is for sparsehistogram branch.

This also takes care of this point in https://github.com/prometheus/prometheus/issues/11171

- [x] Checkpointing works fine with histograms in it

### CODE REVIEW

File: tsdb/wal/checkpoint.go [hunk 1]
```diff
@@ -148,20 +148,21 @@ func Checkpoint(logger log.Logger, w *WAL, from, to int, keep func(id chunks.Hea
...
-		tstones   []tombstones.Stone
-		exemplars []record.RefExemplar
-		metadata  []record.RefMetadata
-		dec       record.Decoder
-		enc       record.Encoder
-		buf       []byte
-		recs      [][]byte
+		series     []record.RefSeries
+		samples    []record.RefSample
+		histograms []record.RefHistogram
```
Comment (beorn7) [inline tsdb/wal/checkpoint.go:153]: As said above, `RefSample` vs. `RefHistogram` feels confusing to me now. It should be more `RefFloatSample` vs. `RefHistogramSample` or something…

I guess renaming in `record` has all kind of implications, but perhaps the local variables here can be more expressive, e.g. `samples` → `floatSamples` and `histograms` → `histogramSamples`. On the other hand, the inconsistency with the `record` package might be confusing. I'm torn…

File: tsdb/wal/checkpoint.go [hunk 2]
```diff
@@ -206,6 +207,24 @@ func Checkpoint(logger log.Logger, w *WAL, from, to int, keep func(id chunks.Hea
 			stats.TotalSamples += len(samples)
 			stats.DroppedSamples += len(samples) - len(repl)
 
+		case record.Histograms:
+			histograms, err = dec.Histograms(rec, histograms)
+			if err != nil {
+				return nil, errors.Wrap(err, "decode samples")
```
Comment (beorn7) [inline tsdb/wal/checkpoint.go:213]: "decode samples" → "decode histograms" (or maybe "decode histogram samples"…).

**Review by beorn7 — CHANGES_REQUESTED** [review]
Thanks. Just minor comments.

I notice that my brain gets really confused by the naming decisions in the `record` package… Perhaps we should at least rename `Histograms` to `HistogramSamples` (and leave the name `Samples` for float samples as is to not change existing usage)… :exploding_head:

File: tsdb/wal/checkpoint.go (hunk 1 above)
Comment (codesome) [reply tsdb/wal/checkpoint.go:153]: At the moment, I will go with calling it histogram samples while keeping the float samples as it is (i.e. calling samples) to start with a less invasive change.

**Pull request merged**